### DISCUSSION TITLE
should take the first title when there are multiple title

### DISCUSSION THREAD

**Discussion by AaronChen0** [projectdiscovery/httpx#430] [General] [answered]
`echo https://blog.thelevelup.com/ | httpx -title`  
Get:      https://blog.thelevelup.com [instagram]
Expect: https://blog.thelevelup.com [Blog | LevelUp | Mobile Ordering, Restaurant Loyalty, Customer Engagement, and CRM |]

When a page has multiple <title>...</title>, the first match probably is the real title.

**Comment by ehsandeep** [comment, marked as answer]
Hi @AaronChen0, there are [multiple](https://github.com/projectdiscovery/httpx/issues/426) variations of titles, and might be not possible to cover all the cases with pre-defined rules, keeping this issue in mind, we added support for custom extractors based on user-defined custom regex, i.e using `-extract-regex`

**Reply by AaronChen0** [reply]
@ehsandeep. This case is not about regex. The first real title is inside the \<head>\</head> tags. Another three \<title> tags are inside the \<body>\</body> tags. So you should match the one inside the \<head>\</head> tags, not the one inside the \<body>\</body> tags. This should be considered a bug, rather than a bizarre regex effort.

**Reply by ehsandeep** [reply]
@AaronChen0 httpx internally uses hardcoded regex only to detect/extract these titles, so regex is what you or we need to fix it, as mentioned we have identified multiple cases that can not satisfy all the cases, but let me know if you have a solution with a better regex to capture all the cases.

https://github.com/projectdiscovery/httpx/blob/bbc8e48ce16aba08c0fb7d88277ddff5f4e201d9/common/httpx/title.go#L15

**Reply by AaronChen0** [reply]
Complex regex would slow down httpx. In this title case, can we just take the first match? Do you have a good reason to take the last match (the current httpx behavior)?

**Reply by ehsandeep** [reply]
@AaronChen0 please go through the multiple issues we have closed regarding title extraction and cases that need to consider to get maximum and valid coverage - https://github.com/projectdiscovery/httpx/issues?q=is%3Aissue+title+is%3Aclosed, at the same time we also understood that it's not possible to capture all the cases, in other word if we update regex to capture new case will result into missing old ones, for the very same reason we added support to let the user define when required.

**Reply by AaronChen0** [reply]
I know nothing about go, but this shouldn't need to change the regex. I want change in after-regex. When the go code have a match list for the regex, why it just take the last one, not the first one. Do you have a reason to do that? I have viewed the other title issues, but none is similar to this one.

**Reply by ehsandeep** [reply]
Hey @AaronChen0, apologies for this, somehow I overlooked the base question/issue you pointed out here, as this is a valid concern I've created GH ticket for this https://github.com/projectdiscovery/httpx/issues/539
